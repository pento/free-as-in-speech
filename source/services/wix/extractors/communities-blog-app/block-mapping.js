/**
 * WordPress dependencies
 */
import { createBlock, serialize } from '@wordpress/blocks';
import { applyFormat, create, toHTMLString } from '@wordpress/rich-text';

/**
 * The definition map for converting Wix blocks to WordPress blocks.
 *
 * The key is the Wix block type.
 *
 * The value is a function that takes the block object, and returns an
 * equivalent WordPress block.
 */
const blockMap = {
	atomic: ( block, { entityMap, ownerSiteMemberId } ) => {
		if ( block.entityRanges.length === 0 ) {
			// eslint-disable-next-line no-console
			console.log( 'Entity block with no entity range', block );
			return false;
		}

		const entity = entityMap[ block.entityRanges[ 0 ].key ];

		switch ( entity.type.toLowerCase() ) {
			case 'wix-draft-plugin-image':
				const imageAttributes = {
					url: `https://static.wixstatic.com/media/${ entity.data.src.file_name }`,
					align: entity.data.config.alignment,
				};

				if ( entity.data.metadata ) {
					imageAttributes.alt = entity.data.metadata.alt;
					imageAttributes.caption = entity.data.metadata.caption;
				}

				if ( entity.data.link ) {
					imageAttributes.href = entity.data.link.url;
					imageAttributes.linkTarget = entity.data.link.target;
					imageAttributes.rel = entity.data.link.rel;
				}

				return createBlock( 'core/image', imageAttributes );

			case 'wix-draft-plugin-video':
				// Uploaded videos should be treated as video blocks.
				if ( entity.data.isCustomVideo ) {
					return createBlock( 'core/video', {
						src: `https://video.wixstatic.com/${ entity.data.src.pathname }`,
						poster: `https://static.wixstatic.com/${ entity.data.src.thumbnail.pathname }`,
					} );
				}

				// All other videos are some form of embed.
				return createBlock( 'core/embed', {
					url: entity.data.src,
					align: entity.data.config.alignment,
				} );

			case 'wix-draft-plugin-gallery':
				const gallerySettings = {};

				gallerySettings.images = entity.data.items.map( ( img ) => ( {
					url: `https://static.wixstatic.com/media/${ img.url }`,
				} ) );

				if ( entity.data.styles.numberOfImagesPerRow > 0 ) {
					gallerySettings.columns =
						entity.data.styles.numberOfImagesPerRow;
				}

				return createBlock( 'core/gallery', gallerySettings );

			case 'wix-draft-plugin-file-upload':
				return createBlock( 'core/file', {
					href: `https://${ ownerSiteMemberId }.usrfiles.com/ugd/${ entity.data.path }`,
					fileName: entity.data.name,
				} );
		}

		// eslint-disable-next-line no-console
		console.log( 'Unknown atomic entity type', entity );
		return false;
	},
	'code-block': ( block, { entityMap } ) => {
		return createBlock( 'core/code', {
			content: formatText( block, entityMap ),
		} );
	},
	blockquote: ( block, { entityMap } ) => {
		return createBlock( 'core/quote', {
			value: '<p>' + formatText( block, entityMap ) + '</p>',
			align: block.data.textAlignment,
		} );
	},
	'header-two': ( block, { entityMap } ) => {
		return createBlock( 'core/heading', {
			content: formatText( block, entityMap ),
			level: 2,
			align: block.data.textAlignment,
		} );
	},
	'header-three': ( block, { entityMap } ) => {
		return createBlock( 'core/heading', {
			content: formatText( block, entityMap ),
			level: 3,
			align: block.data.textAlignment,
		} );
	},
	'ordered-list-item': ( blocks, { entityMap } ) => {
		return createBlock( 'core/list', {
			ordered: true,
			values: blocks
				.map(
					( block ) =>
						'<li>' + formatText( block, { entityMap } ) + '</li>'
				)
				.join( '' ),
		} );
	},
	'unordered-list-item': ( blocks, { entityMap } ) => {
		return createBlock( 'core/list', {
			ordered: false,
			values: blocks
				.map(
					( block ) =>
						'<li>' + formatText( block, { entityMap } ) + '</li>'
				)
				.join( '' ),
		} );
	},
	unstyled: ( block, { entityMap } ) => {
		// Don't transform empty lines into paragraphs.
		if ( ! block.text.trim() ) {
			return false;
		}

		return createBlock( 'core/paragraph', {
			content: formatText( block, entityMap ),
			align: block.data.textAlignment,
		} );
	},
};

/**
 * Given a Wix block, generates a HTML string based upon the block text, the defined
 * inline styles, and any links definied in the entities list.
 *
 * @param {Object} block The Wix block definition.
 * @param {Object} entityMap The post entity map.
 *
 * @return {string} A HTML string generated from the block text.
 */
const formatText = ( block, entityMap ) => {
	if (
		block.inlineStyleRanges.length === 0 &&
		block.entityRanges.length === 0
	) {
		return block.text;
	}

	const formatMap = {
		bold: { type: 'core/bold' },
		italic: { type: 'core/italic' },
		underline: {
			type: 'core/underline',
			attributes: {
				style: 'text-decoration: underline;',
			},
		},
	};

	const richText = block.inlineStyleRanges.reduce( ( workingText, style ) => {
		const styleName = style.style.toLowerCase();

		if ( ! formatMap[ styleName ] ) {
			// eslint-disable-next-line no-console
			console.log( `There's no style registered for '${ styleName }'` );
			return workingText;
		}

		if ( styleName === 'underline' ) {
			// Underline can be unnecessary applied to link text, check if the range
			// of this style matches a link.

			const matchedLink = block.entityRanges.reduce(
				( matched, entityRange ) => {
					if ( matched ) {
						return matched;
					}

					if (
						entityRange.offset !== style.offset ||
						entityRange.length !== style.length
					) {
						return false;
					}

					return (
						entityMap[ entityRange.key ].type.toLowerCase() ===
						'link'
					);
				},
				false
			);

			if ( matchedLink ) {
				return workingText;
			}
		}

		return applyFormat(
			workingText,
			formatMap[ styleName ],
			style.offset,
			style.offset + style.length
		);
	}, create( { text: block.text } ) );

	const entitisedText = block.entityRanges.reduce(
		( workingText, entityRange ) => {
			const entity = entityMap[ entityRange.key ];

			if ( entity.type.toLowerCase() === 'link' ) {
				const linkFormat = {
					type: 'core/link',
					attributes: { url: entity.data.url },
				};

				if ( entity.data.target === '_blank' ) {
					linkFormat.attributes.target = '_blank';
					linkFormat.attributes.rel = 'noreferrer noopener';
				}

				return applyFormat(
					workingText,
					linkFormat,
					entityRange.offset,
					entityRange.offset + entityRange.length
				);
			}

			// eslint-disable-next-line no-console
			console.log( 'Unknown text entity type', entity );
			return workingText;
		},
		richText
	);

	return toHTMLString( { value: entitisedText } );
};

/**
 * Given a list of Wix blocks, convert them to a serialized form of WordPress blocks.
 *
 * @param {Object} wixBlocks The Wix blocks.
 * @param {Object} data      Additional post data.
 *
 * @return {string} The serialized WordPress blocks.
 */
export const serializeWixBlocksToWordPressBlocks = ( wixBlocks, data ) => {
	let listType = '';
	const listBuffer = [];

	return wixBlocks
		.map( ( wixBlock, blockIndex ) => {
			// Wix stores each list item as an individual block, we need to merge them together.
			// Store list blocks in the listBuffer until we're at the last item in the current list,
			// then send that as a single "block".
			if (
				[ 'unordered-list-item', 'ordered-list-item' ].includes(
					wixBlock.type
				)
			) {
				if ( ! listType ) {
					listType = wixBlock.type;
				}
				listBuffer.push( wixBlock );

				if (
					wixBlocks[ blockIndex + 1 ] &&
					wixBlocks[ blockIndex + 1 ].type === listType
				) {
					// The next block is part of the list, so move on to that one.
					return false;
				}

				// We're at the last list item, process the list now.
				const wpBlock = blockMap[ listType ]( listBuffer, data );

				// Clean up the buffer.
				listType = '';
				listBuffer.splice( 0 );

				return wpBlock;
			}

			if ( blockMap[ wixBlock.type ] ) {
				return blockMap[ wixBlock.type ]( wixBlock, data );
			}

			// eslint-disable-next-line no-console
			console.log( 'Unknown Wix block', {
				block: wixBlock,
				data,
			} );
			return false;
		} )
		.filter( ( blockContent ) => blockContent !== false )
		.map( ( wpBlock ) => serialize( wpBlock ) )
		.join( '\n\n' );
};
