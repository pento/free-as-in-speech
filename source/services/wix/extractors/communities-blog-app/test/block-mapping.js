/**
 * WordPress dependencies
 */
import { registerCoreBlocks } from '@wordpress/block-library';
import '@wordpress/format-library';

/**
 * Internal dependencies
 */
import { serializeWixBlocksToWordPressBlocks } from '../block-mapping';

const testData = {
	entityMap: {
		0: {
			type: 'LINK',
			mutability: 'MUTABLE',
			data: {
				url: 'https://wordpress.org/',
				target: '_blank',
				rel: 'noopener',
			},
		},
		1: {
			type: 'wix-draft-plugin-image',
			data: {
				src: {
					file_name: 'image1.png',
				},
				config: {
					alignment: 'center',
				},
				metadata: {
					alt: 'image1 alt text',
					caption: 'image1 caption',
				},
				link: {
					url: 'https://wordpress.org/',
					target: '_blank',
					rel: 'noopener',
				},
			},
		},
		2: {
			type: 'wix-draft-plugin-video',
			data: {
				isCustomVideo: true,
				src: {
					pathname: 'video1.mp4',
					thumbnail: {
						pathname: 'media/video1_thumb.jpg',
					},
				},
			},
		},
		3: {
			type: 'wix-draft-plugin-video',
			data: {
				src: 'https://www.youtube.com/watch?v=RWf83UX4vKs',
				config: {
					alignment: 'right',
				},
			},
		},
		4: {
			type: 'wix-draft-plugin-gallery',
			data: {
				items: [
					{ url: 'image2.png' },
					{ url: 'image3.png' },
					{ url: 'image4.png' },
					{ url: 'image5.png' },
					{ url: 'image6.png' },
					{ url: 'image7.png' },
					{ url: 'image8.png' },
				],
				styles: {
					numberOfImagesPerRow: 4,
				},
			},
		},
	},
};

describe( 'Block Mapping', () => {
	beforeAll( () => {
		registerCoreBlocks();
	} );

	test( 'serializeWixBlocksToWordPressBlocks is a function callback function', () => {
		expect( typeof serializeWixBlocksToWordPressBlocks ).toBe( 'function' );
	} );

	test( 'serializeWixBlocksToWordPressBlocks returns an empty string when passed no blocks', () => {
		expect( serializeWixBlocksToWordPressBlocks( [], testData ) ).toEqual(
			''
		);
	} );

	test( 'serializeWixBlocksToWordPressBlocks handles "unstyled" blocks as paragraphs', () => {
		const blocks = [
			{
				type: 'unstyled',
				text: 'This is a paragraph',
				inlineStyleRanges: [],
				entityRanges: [],
				data: {},
			},
			{
				type: 'unstyled',
				text: 'This is another paragraph',
				inlineStyleRanges: [],
				entityRanges: [],
				data: {},
			},
		];

		const expected = `<!-- wp:paragraph -->
<p>This is a paragraph</p>
<!-- /wp:paragraph -->

<!-- wp:paragraph -->
<p>This is another paragraph</p>
<!-- /wp:paragraph -->`;

		expect(
			serializeWixBlocksToWordPressBlocks( blocks, testData )
		).toEqual( expected );
	} );

	test( 'serializeWixBlocksToWordPressBlocks handles heading blocks', () => {
		const blocks = [
			{
				type: 'header-two',
				text: 'This is a h2',
				inlineStyleRanges: [],
				entityRanges: [],
				data: {},
			},
			{
				type: 'header-three',
				text: 'This is a h3',
				inlineStyleRanges: [],
				entityRanges: [],
				data: {},
			},
		];

		const expected = `<!-- wp:heading -->
<h2>This is a h2</h2>
<!-- /wp:heading -->

<!-- wp:heading {"level":3} -->
<h3>This is a h3</h3>
<!-- /wp:heading -->`;

		expect(
			serializeWixBlocksToWordPressBlocks( blocks, testData )
		).toEqual( expected );
	} );

	test( 'serializeWixBlocksToWordPressBlocks handles code blocks', () => {
		const blocks = [
			{
				type: 'code-block',
				text:
					"This is some code\nIt's possibly the best code I've ever written\nmaybe.",
				inlineStyleRanges: [],
				entityRanges: [],
				data: {},
			},
		];

		const expected = `<!-- wp:code -->
<pre class="wp-block-code"><code>This is some code
It's possibly the best code I've ever written
maybe.</code></pre>
<!-- /wp:code -->`;

		expect(
			serializeWixBlocksToWordPressBlocks( blocks, testData )
		).toEqual( expected );
	} );

	test( 'serializeWixBlocksToWordPressBlocks handles quote blocks', () => {
		const blocks = [
			{
				type: 'blockquote',
				text: 'Use the source, Luke.',
				inlineStyleRanges: [],
				entityRanges: [],
				data: {},
			},
		];

		const expected = `<!-- wp:quote -->
<blockquote class="wp-block-quote"><p>Use the source, Luke.</p></blockquote>
<!-- /wp:quote -->`;

		expect(
			serializeWixBlocksToWordPressBlocks( blocks, testData )
		).toEqual( expected );
	} );

	test( 'serializeWixBlocksToWordPressBlocks handles list items', () => {
		const blocks = [
			{
				type: 'ordered-list-item',
				text: 'Number one',
				inlineStyleRanges: [],
				entityRanges: [],
				data: {},
			},
			{
				type: 'ordered-list-item',
				text: 'Number two',
				inlineStyleRanges: [],
				entityRanges: [],
				data: {},
			},
			{
				type: 'ordered-list-item',
				text: 'Number three',
				inlineStyleRanges: [],
				entityRanges: [],
				data: {},
			},
			{
				type: 'unordered-list-item',
				text: 'First',
				inlineStyleRanges: [],
				entityRanges: [],
				data: {},
			},
			{
				type: 'unordered-list-item',
				text: 'Second',
				inlineStyleRanges: [],
				entityRanges: [],
				data: {},
			},
			{
				type: 'ordered-list-item',
				text: 'Number one, second edition',
				inlineStyleRanges: [],
				entityRanges: [],
				data: {},
			},
		];

		const expected = `<!-- wp:list {"ordered":true} -->
<ol><li>Number one</li><li>Number two</li><li>Number three</li></ol>
<!-- /wp:list -->

<!-- wp:list -->
<ul><li>First</li><li>Second</li></ul>
<!-- /wp:list -->

<!-- wp:list {"ordered":true} -->
<ol><li>Number one, second edition</li></ol>
<!-- /wp:list -->`;

		expect(
			serializeWixBlocksToWordPressBlocks( blocks, testData )
		).toEqual( expected );
	} );

	test( 'serializeWixBlocksToWordPressBlocks handles links in text', () => {
		const blocks = [
			{
				type: 'unstyled',
				text: 'This is a link.',
				inlineStyleRanges: [],
				entityRanges: [
					{
						offset: 8,
						length: 6,
						key: 0,
					},
				],
				data: {},
			},
		];

		const expected = `<!-- wp:paragraph -->
<p>This is <a href="https://wordpress.org/" target="_blank" rel="noreferrer noopener">a link</a>.</p>
<!-- /wp:paragraph -->`;

		expect(
			serializeWixBlocksToWordPressBlocks( blocks, testData )
		).toEqual( expected );
	} );

	test( 'serializeWixBlocksToWordPressBlocks handles formatting in text', () => {
		const blocks = [
			{
				type: 'unstyled',
				text: 'This is bold, italic, underlined, and a combination.',
				inlineStyleRanges: [
					{
						style: 'BOLD',
						offset: 8,
						length: 4,
					},
					{
						style: 'ITALIC',
						offset: 14,
						length: 6,
					},
					{
						style: 'UNDERLINE',
						offset: 22,
						length: 10,
					},
					{
						style: 'BOLD',
						offset: 34,
						length: 17,
					},
					{
						style: 'UNDERLINE',
						offset: 34,
						length: 3,
					},
					{
						style: 'ITALIC',
						offset: 38,
						length: 13,
					},
				],
				entityRanges: [],
				data: {},
			},
		];

		const expected = `<!-- wp:paragraph -->
<p>This is <strong>bold</strong>, <em>italic</em>, <span style="text-decoration: underline;">underlined</span>, <strong><span style="text-decoration: underline;">and</span> <em>a combination</em></strong>.</p>
<!-- /wp:paragraph -->`;

		expect(
			serializeWixBlocksToWordPressBlocks( blocks, testData )
		).toEqual( expected );
	} );

	test( 'serializeWixBlocksToWordPressBlocks ignores underline formatting that matches a link', () => {
		const blocks = [
			{
				type: 'unstyled',
				text: 'This is a link.',
				inlineStyleRanges: [
					{
						style: 'UNDERLINE',
						offset: 8,
						length: 6,
					},
				],
				entityRanges: [
					{
						offset: 8,
						length: 6,
						key: 0,
					},
				],
				data: {},
			},
		];

		const expected = `<!-- wp:paragraph -->
<p>This is <a href="https://wordpress.org/" target="_blank" rel="noreferrer noopener">a link</a>.</p>
<!-- /wp:paragraph -->`;

		expect(
			serializeWixBlocksToWordPressBlocks( blocks, testData )
		).toEqual( expected );
	} );

	test( 'serializeWixBlocksToWordPressBlocks handles single images', () => {
		const blocks = [
			{
				type: 'atomic',
				text: '',
				inlineStyleRanges: [],
				entityRanges: [
					{
						offset: 0,
						length: 0,
						key: 1,
					},
				],
				data: {},
			},
		];

		const expected = `<!-- wp:image {"align":"center"} -->
<div class="wp-block-image"><figure class="aligncenter"><a href="https://wordpress.org/" target="_blank" rel="noopener"><img src="https://static.wixstatic.com/media/image1.png" alt="image1 alt text"/></a><figcaption>image1 caption</figcaption></figure></div>
<!-- /wp:image -->`;

		expect(
			serializeWixBlocksToWordPressBlocks( blocks, testData )
		).toEqual( expected );
	} );

	test( 'serializeWixBlocksToWordPressBlocks handles uploaded videos', () => {
		const blocks = [
			{
				type: 'atomic',
				text: '',
				inlineStyleRanges: [],
				entityRanges: [
					{
						offset: 0,
						length: 0,
						key: 2,
					},
				],
				data: {},
			},
		];

		const expected = `<!-- wp:video -->
<figure class="wp-block-video"><video controls poster="https://static.wixstatic.com/media/video1_thumb.jpg" src="https://video.wixstatic.com/video1.mp4"></video></figure>
<!-- /wp:video -->`;

		expect(
			serializeWixBlocksToWordPressBlocks( blocks, testData )
		).toEqual( expected );
	} );

	test( 'serializeWixBlocksToWordPressBlocks handles embedded videos', () => {
		const blocks = [
			{
				type: 'atomic',
				text: '',
				inlineStyleRanges: [],
				entityRanges: [
					{
						offset: 0,
						length: 0,
						key: 3,
					},
				],
				data: {},
			},
		];

		const expected = `<!-- wp:embed {"url":"https://www.youtube.com/watch?v=RWf83UX4vKs","align":"right"} -->
<figure class="wp-block-embed alignright"><div class="wp-block-embed__wrapper">
https://www.youtube.com/watch?v=RWf83UX4vKs
</div></figure>
<!-- /wp:embed -->`;

		expect(
			serializeWixBlocksToWordPressBlocks( blocks, testData )
		).toEqual( expected );
	} );

	test( 'serializeWixBlocksToWordPressBlocks handles image galleries', () => {
		const blocks = [
			{
				type: 'atomic',
				text: '',
				inlineStyleRanges: [],
				entityRanges: [
					{
						offset: 0,
						length: 0,
						key: 4,
					},
				],
				data: {},
			},
		];

		const expected = `<!-- wp:gallery {"columns":4} -->
<figure class="wp-block-gallery columns-4 is-cropped"><ul class="blocks-gallery-grid"><li class="blocks-gallery-item"><figure><img src="https://static.wixstatic.com/media/image2.png"/></figure></li><li class="blocks-gallery-item"><figure><img src="https://static.wixstatic.com/media/image3.png"/></figure></li><li class="blocks-gallery-item"><figure><img src="https://static.wixstatic.com/media/image4.png"/></figure></li><li class="blocks-gallery-item"><figure><img src="https://static.wixstatic.com/media/image5.png"/></figure></li><li class="blocks-gallery-item"><figure><img src="https://static.wixstatic.com/media/image6.png"/></figure></li><li class="blocks-gallery-item"><figure><img src="https://static.wixstatic.com/media/image7.png"/></figure></li><li class="blocks-gallery-item"><figure><img src="https://static.wixstatic.com/media/image8.png"/></figure></li></ul></figure>
<!-- /wp:gallery -->`;

		expect(
			serializeWixBlocksToWordPressBlocks( blocks, testData )
		).toEqual( expected );
	} );
} );
