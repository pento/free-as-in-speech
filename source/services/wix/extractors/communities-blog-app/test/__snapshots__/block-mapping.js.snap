// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Block Mapping serializeWixBlocksToWordPressBlocks handles "unstyled" blocks as paragraphs 1`] = `
"<!-- wp:paragraph -->
<p>This is a paragraph</p>
<!-- /wp:paragraph -->

<!-- wp:paragraph -->
<p>This is another paragraph</p>
<!-- /wp:paragraph -->"
`;

exports[`Block Mapping serializeWixBlocksToWordPressBlocks handles code blocks 1`] = `
"<!-- wp:code -->
<pre class=\\"wp-block-code\\"><code>This is some code
It's possibly the best code I've ever written
maybe.</code></pre>
<!-- /wp:code -->"
`;

exports[`Block Mapping serializeWixBlocksToWordPressBlocks handles embedded videos 1`] = `
"<!-- wp:embed {\\"url\\":\\"https://www.youtube.com/watch?v=RWf83UX4vKs\\",\\"align\\":\\"right\\"} -->
<figure class=\\"wp-block-embed alignright\\"><div class=\\"wp-block-embed__wrapper\\">
https://www.youtube.com/watch?v=RWf83UX4vKs
</div></figure>
<!-- /wp:embed -->"
`;

exports[`Block Mapping serializeWixBlocksToWordPressBlocks handles formatting in text 1`] = `
"<!-- wp:paragraph -->
<p>This is <strong>bold</strong>, <em>italic</em>, <span style=\\"text-decoration: underline;\\">underlined</span>, <strong><span style=\\"text-decoration: underline;\\">and</span> <em>a combination</em></strong>.</p>
<!-- /wp:paragraph -->"
`;

exports[`Block Mapping serializeWixBlocksToWordPressBlocks handles heading blocks 1`] = `
"<!-- wp:heading -->
<h2>This is a h2</h2>
<!-- /wp:heading -->

<!-- wp:heading {\\"level\\":3} -->
<h3>This is a h3</h3>
<!-- /wp:heading -->"
`;

exports[`Block Mapping serializeWixBlocksToWordPressBlocks handles image galleries 1`] = `
"<!-- wp:gallery {\\"columns\\":4} -->
<figure class=\\"wp-block-gallery columns-4 is-cropped\\"><ul class=\\"blocks-gallery-grid\\"><li class=\\"blocks-gallery-item\\"><figure><img src=\\"https://static.wixstatic.com/media/image2.png\\"/></figure></li><li class=\\"blocks-gallery-item\\"><figure><img src=\\"https://static.wixstatic.com/media/image3.png\\"/></figure></li><li class=\\"blocks-gallery-item\\"><figure><img src=\\"https://static.wixstatic.com/media/image4.png\\"/></figure></li><li class=\\"blocks-gallery-item\\"><figure><img src=\\"https://static.wixstatic.com/media/image5.png\\"/></figure></li><li class=\\"blocks-gallery-item\\"><figure><img src=\\"https://static.wixstatic.com/media/image6.png\\"/></figure></li><li class=\\"blocks-gallery-item\\"><figure><img src=\\"https://static.wixstatic.com/media/image7.png\\"/></figure></li><li class=\\"blocks-gallery-item\\"><figure><img src=\\"https://static.wixstatic.com/media/image8.png\\"/></figure></li></ul></figure>
<!-- /wp:gallery -->"
`;

exports[`Block Mapping serializeWixBlocksToWordPressBlocks handles links in text 1`] = `
"<!-- wp:paragraph -->
<p>This is <a href=\\"https://wordpress.org/\\" target=\\"_blank\\" rel=\\"noreferrer noopener\\">a link</a>.</p>
<!-- /wp:paragraph -->"
`;

exports[`Block Mapping serializeWixBlocksToWordPressBlocks handles list items 1`] = `
"<!-- wp:list {\\"ordered\\":true} -->
<ol><li>Number one</li><li>Number two</li><li>Number three</li></ol>
<!-- /wp:list -->

<!-- wp:list -->
<ul><li>First</li><li>Second</li></ul>
<!-- /wp:list -->

<!-- wp:list {\\"ordered\\":true} -->
<ol><li>Number one, second edition</li></ol>
<!-- /wp:list -->"
`;

exports[`Block Mapping serializeWixBlocksToWordPressBlocks handles quote blocks 1`] = `
"<!-- wp:quote -->
<blockquote class=\\"wp-block-quote\\"><p>Use the source, Luke.</p></blockquote>
<!-- /wp:quote -->"
`;

exports[`Block Mapping serializeWixBlocksToWordPressBlocks handles single images 1`] = `
"<!-- wp:image {\\"align\\":\\"center\\"} -->
<div class=\\"wp-block-image\\"><figure class=\\"aligncenter\\"><a href=\\"https://wordpress.org/\\" target=\\"_blank\\" rel=\\"noopener\\"><img src=\\"https://static.wixstatic.com/media/image1.png\\" alt=\\"image1 alt text\\"/></a><figcaption>image1 caption</figcaption></figure></div>
<!-- /wp:image -->"
`;

exports[`Block Mapping serializeWixBlocksToWordPressBlocks handles uploaded videos 1`] = `
"<!-- wp:video -->
<figure class=\\"wp-block-video\\"><video controls poster=\\"https://static.wixstatic.com/media/video1_thumb.jpg\\" src=\\"https://video.wixstatic.com/video1.mp4\\"></video></figure>
<!-- /wp:video -->"
`;

exports[`Block Mapping serializeWixBlocksToWordPressBlocks ignores underline formatting that matches a link 1`] = `
"<!-- wp:paragraph -->
<p>This is <a href=\\"https://wordpress.org/\\" target=\\"_blank\\" rel=\\"noreferrer noopener\\">a link</a>.</p>
<!-- /wp:paragraph -->"
`;

exports[`Block Mapping serializeWixBlocksToWordPressBlocks returns an empty string when passed no blocks 1`] = `""`;
